plugin/methodWrappers.go

```diff
@@ -0,0 +1,139 @@
+//
+// NOTE: <plugin_name> is really <plugin_cname>. However since <plugin_name>
+// can never contain a '/', <plugin_cname> reduces to <plugin_name>.
+func ID(e Entry) string {
+	if e.id() == "" {
+		msg := fmt.Sprintf("plugin.ID: entry %v (cname %v) has no ID", e.name(), CName(e))
+		panic(msg)
+	}
+	return e.id()
+}
+
+// Attributes returns the entry's attributes. If size is unknown, it will check whether the entry
+// has locally cached content and if so set that for the size.
+func Attributes(e Entry) EntryAttributes {
+	// Sometimes an entry doesn't know its size unless it's already downloaded some content. Having
+	// to download content makes list slow, and is a burden for external plugin developers. Check if
+	// we already know the size. If not, FUSE will use a reasonable default so tools don't ignore it.
+	attr := e.attributes()
+	if !attr.HasSize() && cache != nil {
+		// We have no way to preserve this on the entry, and it likely wouldn't help because we often
+		// recreate the entry to ensure we have an accurate representation. So when the cache expires
+		// we revert to stating the size is unknown until the next read operation.
+		if val, _ := cache.Get(defaultOpCodeToNameMap[OpenOp], e.id()); val != nil {
+			rdr := val.(SizedReader)
+			attr.SetSize(uint64(rdr.Size()))
+		}
+	}
+	return attr
+}
+
+// IsPrefetched returns whether an entry has data that was added during creation that it would
+// like to have updated.
+func IsPrefetched(e Entry) bool {
+	return e.isPrefetched()
+}
+
+// Schema returns the entry's schema.
+func Schema(e Entry) (*EntrySchema, error) {
+	return schema(e)
+}
+
+// List lists the parent's children. It returns a map of <entry_cname> => <entry_object>
+// to optimize querying a specific entry.
+//
+// Note that List's results could be cached.
+func List(ctx context.Context, p Parent) (map[string]Entry, error) {
+	return cachedList(ctx, p)
+}
+
+// Open reads the entry's content. Note that Open's results could be cached. Thus, when
+// using the reader returned by this method, use idempotent read operations such as ReadAt
+// or wrap it in a SectionReader. Using Read operations on the cached reader will change it
+// and make subsequent uses of the cached reader invalid.
+//
+// TODO: Could we change this to Read? E.g. plugin.Read.
+func Open(ctx context.Context, r Readable) (SizedReader, error) {
+	return cachedOpen(ctx, r)
+}
+
+// Metadata returns the entry's metadata. Note that Metadata's results could be cached.
+func Metadata(ctx context.Context, e Entry) (JSONObject, error) {
+	return cachedMetadata(ctx, e)
+}
+
+// Exec execs the command on the given entry.
+func Exec(ctx context.Context, e Execable, cmd string, args []string, opts ExecOptions) (ExecCommand, error) {
+	return e.Exec(ctx, cmd, args, opts)
+}
+
+// Stream streams the entry's content for updates.
+func Stream(ctx context.Context, s Streamable) (io.ReadCloser, error) {
+	return s.Stream(ctx)
+}
```

Comment:
I like this better because now I can add `Delete` and centralize the cache-related cleanup involved when deleting an entry.